documents/repository.go

```diff
@@ -74,12 +115,117 @@ func (r *repo) Get(accountID, id []byte) (Model, error) {
 	key := r.getKey(accountID, id)
-	return r.db.Create(key, model)
+	if err := r.db.Create(key, model); err != nil {
+		return err
+	}
+
+	return r.updateLatestIndex(accountID, model)
 }
 
 // Update strictly updates the model.
 // Will error out when the model doesn't exist in the DB.
 func (r *repo) Update(accountID, id []byte, model Model) error {
 	key := r.getKey(accountID, id)
-	return r.db.Update(key, model)
+	if err := r.db.Update(key, model); err != nil {
+		return err
+	}
+
+	return r.updateLatestIndex(accountID, model)
+}
+
+// GetLatest returns thee latest version of the document.
+func (r *repo) GetLatest(accountID, docID []byte) (Model, error) {
+	key := r.getLatestKey(accountID, docID)
+	lv, err := r.getLatest(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return r.Get(accountID, lv.CurrentVersion)
+}
+
+func (r *repo) getLatest(key []byte) (*latestVersion, error) {
+	val, err := r.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	lv, ok := val.(*latestVersion)
+	if ok {
+		return lv, nil
+	}
+
+	// delete key val if the type mismatches
+	err = r.db.Delete(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, ErrDocumentNotFound
+}
+
+// getLatestKey constructs the key to the latest version of the document.
+// Note: DocumentIdentifier needs to be passed here not the versionID.
+func (r *repo) getLatestKey(accountID, docID []byte) []byte {
+	hexKey := hexutil.Encode(append(accountID, docID...))
+	return append([]byte(LatestPrefix), []byte(hexKey)...)
+}
+
+// storeLatestIndex stores the latestVersion to db.
+// If update is true, it is assumed that index is overwritten
+// else, index is created first time.
+func (r *repo) storeLatestIndex(key []byte, model Model, update bool) error {
+	lv := &latestVersion{
+		CurrentVersion: model.CurrentVersion(),
+		NextVersion:    model.NextVersion(),
+	}
+
+	tm, err := model.Timestamp()
+	if err != nil {
+		// we will update this actual value when available
+		tm = time.Now().UTC()
+	}
+	lv.Timestamp = tm
+
+	if update {
+		return r.db.Update(key, lv)
+	}
+
+	return r.db.Create(key, lv)
+}
+
+// updateLatestIndex updates the latest version index.
+// We check if the latest index is present for a model.
+// If not found, create a latest index and return.
+// Note: anchor timestamp is not available immediately, so don't error out if the timestamp is empty
+// If found, check if the next version matches the current version of the passed model.
+// If matches, update the timestamp of anchor and return.
+// If not matches, check the model timestamp is greater than stored timestamp.
+// If greater update the latestVersion and return
+// If not, skip update and return.
+func (r *repo) updateLatestIndex(accID []byte, model Model) error {
+	key := r.getLatestKey(accID, model.ID())
+	lv, err := r.getLatest(key)
+	if err != nil {
+		// no index is created yet. create one
+		return r.storeLatestIndex(key, model, false)
+	}
+
+	if bytes.Equal(lv.NextVersion, model.CurrentVersion()) {
```

Comment:
If the timestamp fails, we would fall with time.Now(). I would like to keep this check since it will be more determinstic

Reply:
In the case you jump many versions, let's say you have version 1 and trying to update version 3, this check will not be executed, and then the below code will be triggered. In which cases there will not be any timestamp?

Reply:
Sure, when you jump versions this wont be triggered and hence we have time stamp check for it. But when its is the next logical version, this will be triggered. Timestamp will not be there when the document is created and immediately saved

Reply:
> Sure, when you jump versions this wont be triggered and hence we have time stamp check for it. But when its is the next logical version, this will be triggered.

I guess that is a question on how to act if a next version document timestamp is older than the existing one, then should we store it? 
> Timestamp will not be there when the document is created and immediately saved

Then it will trigger be ok setting the now timestamp.

Reply:
> I guess that is a question on how to act if a next version document timestamp is older than the existing one, then should we store it?

We should not worry about it here since the Validators wouldn't let it pass through before this step.

> Then it will trigger be ok setting the now timestamp.

I think we should only fallback to this when we are not sure about the version. If we are sure about it, then why not take that route?